hazelcast/src/main/java/com/hazelcast/concurrent/flakeidgen/FlakeIdGeneratorProxy.java

```diff
@@ -0,0 +1,198 @@
+        // node IDs are assigned from 0 again.
+        // It won't be possible to generate IDs through member HZ instances, however clients ignore
+        // members which threw FlakeIdNodeIdOutOfRangeException, so they'll work as long as there is
+        // at least one member with node ID in the cluster.
+        if ((nodeId & -1 << BITS_NODE_ID) != 0) {
+            this.nodeId = -1;
+            logger.severe("Node ID is out of range (" + nodeId + "), this member won't be able to generate IDs. "
+                    + "Cluster restart is recommended.");
+        } else {
+            this.nodeId = nodeId;
+        }
+
+        FlakeIdGeneratorConfig config = getNodeEngine().getConfig().findFlakeIdGeneratorConfig(getName());
+        batcher = new AutoBatcher(config.getPrefetchCount(), config.getPrefetchValidity(), new IFunction<Integer, IdBatch>() {
+            @Override
+            public IdBatch apply(Integer batchSize) {
+                return newIdBatch(batchSize);
+            }
+        });
+    }
+
+    @Override
+    public long newId() {
+        return batcher.newId();
+    }
+
+    @Override
+    public IdBatch newIdBatch(int batchSize) {
+        while (true) {
+            NewIdBatchOperation op = new NewIdBatchOperation(name, batchSize);
+            Member member = getRandomMember();
+            InternalCompletableFuture<Long> future = getNodeEngine().getOperationService()
+                                                                    .invokeOnTarget(getServiceName(), op, member.getAddress());
+            try {
+                long base = future.join();
+                return new IdBatch(base, 1 << BITS_NODE_ID, batchSize);
+            } catch (FlakeIdNodeIdOutOfRangeException e) {
+                outOfRangeMembers.add(member.getUuid());
+                randomMember = null;
+            }
+        }
+    }
+
+    long newIdBaseLocal(int batchSize) {
+        return newIdBaseLocal(Clock.currentTimeMillis(), batchSize);
+    }
+
+    /**
+     * The layout of the ID is as follows (starting from most significant bits):<ul>
+     *     <li>42 bits timestamp
+     *     <li>6 bits sequence
+     *     <li>16 bits node ID
+     * </ul>
+     *
+     * This order is important: timestamp must be first to keep IDs ordered. Sequence must be second for
+     * implementation reasons (it's included in {@link #generatedValue}). Node is just an appendix to make
+     * IDs unique.
+     *
+     * @param now Current time (currentTimeMillis() normally or other value in tests)
+     * @return First ID for the batch.
+     */
+    long newIdBaseLocal(long now, int batchSize) {
+        checkPositive(batchSize, "batchSize");
+        if (nodeId < 0) {
+            logger.info("Throwing FlakeIdNodeIdOutOfRangeException");
+            throw new FlakeIdNodeIdOutOfRangeException("NodeID overflow, this member cannot generate IDs");
+        }
+        logger.info("Generating " + batchSize + " IDs");
```

Comment:
I would rather not log at `INFO` level for each batch. Maybe `fine` or `finest`

Reply:
I intended to remove this, thanks for spotting it...